parlai/mturk/core/dev/react_server/server/server.js

```diff
@@ -31,17 +31,289 @@ nunjucks.configure(task_directory_name, {
+    let new_conversation_id =
+      state_change['conversation_id'] || this.conversation_id;
+    let new_status = state_change['agent_status'] || this.status;
+    let done_text = state_change['done_text'] || this.done_text;
+    let agent_id = state_change['agent_id'] || this.agent_id;
+
+    let update_packet = {};
+    let needs_update = false;
+    if (new_conversation_id != this.conversation_id) {
+      needs_update = true;
+      update_packet['conversation_id'] = new_conversation_id;
+      this.conversation_id = new_conversation_id;
+      this._init_conversation(new_conversation_id);
+    }
+
+    if (new_status != this.status) {
+      needs_update = true;
+      update_packet['agent_status'] = new_status;
+      this.status = new_status;
+
+      if (is_final_status(new_status)) {
+        update_packet['final'] = true;
+        this._cleanup_state();
+      }
+
+      if (new_status == STATUS_STATIC) {
+        // Status static needs special cleanup to keep state but stop tracking
+        let sendable_messages = this.get_sendable_messages()
+        if (sendable_messages.length > 0) {
+          let packet = {'messages': sendable_messages};
+          _send_message(this.connection_id, MESSAGE_BATCH, packet);
+        }
+        if (active_connections.has(this.connection_id)) {
+          active_connections.delete(this.connection_id);
+        }
+      }
+    }
+
+    if (done_text != this.done_text) {
+      needs_update = true;
+      update_packet['done_text'] = done_text;
+      this.done_text = done_text;
+    }
+
+    if (agent_id != this.agent_id) {
+      needs_update = true;
+      update_packet['agent_id'] = agent_id;
+      this.agent_id = agent_id;
+    }
+
+    if (needs_update) {
+      return update_packet;
+    } else {
+      return null;
+    }
+  }
+
+  new_message(msg) {
+    if (this.received_message_ids.has(msg['id'])) {
+      return;  // This message has already been added
+    }
+
+    if (is_final_status(this.status)) {
+      // There's no message queue to deliver this message to anymore
+      console.log("msg queue was already cleaned up");
+      console.log(msg);
+      return;
+    }
+
+    // rare new_message called before set_status race condition
+    if (this.previous_messages[msg['conversation_id']] === undefined) {
+      this._init_conversation(msg['conversation_id']);
+    }
+
+    this.unsent_messages[msg['conversation_id']].push(msg);
+    this.previous_messages[msg['conversation_id']].push(msg);
+  }
+
+  get_sendable_messages() {
+    if (this.conversation_id == null) {
+      return [];
+    }
+    let unsent_messages = this.unsent_messages[this.conversation_id] || [];
+    let return_messages = [];
+    while (unsent_messages.length > 0) {
+      return_messages.push(unsent_messages.shift())
+    }
+    return return_messages;
+  }
+
+  add_sent_message(sent_message) {
+    this.previous_messages[sent_message['conversation_id']].push(sent_message);
```

Comment:
will there be any circumstance where `sent_message['conversation_id']` is not in `this.previous_messages`?

Reply:
Under the current architecture no, this could only occur if the MTurkManager was trying to send a message to a conversation that it didn't produce a conversation_id change for yet (ie, someone manually changed an agent's `conversation_id` directly without using access functions), however there's no real harm in doing a check and calling `_init_conversation` here before doing the dereference.